### PR TITLE
Forecasting Data Format

### PR DESCRIPTION
**Pull request by pfistfl**
See [RFC](https://github.com/mlr-org/mlr3forecasting/blob/databformat/attic/data_backend_forecasting.md)

### CODE REVIEW

**Comment by SteveBronder** [comment]
Link to md view
https://github.com/mlr-org/mlr3forecasting/blob/databformat/attic/data_backend_forecasting.md

File: attic/data_backend_forecasting.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+  3: mdeaths 1974-03-01  1877
+  4: mdeaths 1974-04-01  1877
+  5: mdeaths 1974-05-01  1492
+ ---                         
+140: fdeaths 1979-08-01   379
+141: fdeaths 1979-09-01   393
+142: fdeaths 1979-10-01   411
+143: fdeaths 1979-11-01   487
+144: fdeaths 1979-12-01   574
+```
```
Comment (SteveBronder) [inline attic/data_backend_forecasting.md]: I like the long form, though then we I fear we will spend a lot of time casting to wide for instance with `auto.arima`'s `xreg` it wants a matrix of variables of the same length as y.

Comment (SteveBronder) [reply attic/data_backend_forecasting.md]: Though the long form is very nice for preprocessing, above can take apply lags and other univariate preprocessing nicely

Comment (mllg) [reply attic/data_backend_forecasting.md]: :+1: for long format. Casting to wide should be inexpensive compared to arima.

Comment (SteveBronder) [reply attic/data_backend_forecasting.md]: Word aight I'm cool with it then!

**Comment by SteveBronder** [comment]
Does merging this spec by Saturday work for everyone? Want to make sure this keeps rolling but also give time for people to have a read and ask Qs

**Comment by SteveBronder** [comment]
Reference level impl needs to be fleshed out, dunno if we need a meta func to interact with indices etc.

**Comment by pfistfl** [comment]
@mllg 
Two open questions: 

- `tsbox` looks quite nice and dependencies are AFAIK not bad. Should we use it as a converter from/to different formats? 

- Does the current format need a specific DataBackend, or just a different Task Type?

**Comment by EmilRehnberg** [comment]
Hi guys, sorry for an unrelated question. 
What's the idea behind the `attic` folder? Is it for specifying the outline of the package?

**Comment by pfistfl** [comment]
Hey, the idea was to have something similar to [rust rfc](https://github.com/rust-lang/rfcs) to discuss the design speccs of the package.
Did not know where to put it, and `attic` is sort of where we often stash stuff.

**Comment by annakrystalli** [comment]
Hi all, Just catching up with all the progress.

With respect to data formats, I'd also like to throw in [`tsibble`](https://github.com/tidyverts/tsibble) which is a dependency of `tsbox`.
My understanding is, most common data formats can be converted to a `tsibble`, and that it is overall more flexible.

By default, tsibbles are converted to long format but can easily be converetd to wide instead.
``` r
z <- ts(matrix(rnorm(300), 100, 3), start = c(1961, 1), frequency = 12)
head(z)
#>        Series 1    Series 2    Series 3
#> [1,] -0.1197573 -1.09404086 -1.61006979
#> [2,]  0.1283231 -0.05140625 -0.11975698
#> [3,]  0.3081317 -0.83054010  0.10305690
#> [4,]  0.4491850 -1.36209585  0.97892967
#> [5,] -0.2036575 -0.17698463  0.79885819
#> [6,]  1.9458651 -1.35569709  0.06211394
# long format
tsibble::as_tsibble(z)
#> # A tsibble: 300 x 3 [1M]
#> # Key:       key [3]
#>       index key       value
#>       <mth> <chr>     <dbl>
#>  1 1961 Jan Series 1 -0.120
#>  2 1961 Feb Series 1  0.128
#>  3 1961 Mar Series 1  0.308
#>  4 1961 Apr Series 1  0.449
#>  5 1961 May Series 1 -0.204
#>  6 1961 Jun Series 1  1.95 
#>  7 1961 Jul Series 1 -0.655
#>  8 1961 Aug Series 1  0.918
#>  9 1961 Sep Series 1  0.251
#> 10 1961 Oct Series 1  1.89 
#> # … with 290 more rows
# wide format
tsibble::as_tsibble(z, pivot_longer = FALSE)
#> # A tsibble: 100 x 4 [1M]
#>       index `Series 1` `Series 2` `Series 3`
#>       <mth>      <dbl>      <dbl>      <dbl>
#>  1 1961 Jan     -0.120    -1.09      -1.61  
#>  2 1961 Feb      0.128    -0.0514    -0.120 
#>  3 1961 Mar      0.308    -0.831      0.103 
#>  4 1961 Apr      0.449    -1.36       0.979 
#>  5 1961 May     -0.204    -0.177      0.799 
#>  6 1961 Jun      1.95     -1.36       0.0621
#>  7 1961 Jul     -0.655    -0.397     -0.933 
#>  8 1961 Aug      0.918     1.62      -0.755 
#>  9 1961 Sep      0.251     0.295     -0.0502
#> 10 1961 Oct      1.89     -0.650     -1.08  
#> # … with 90 more rows
```

<sup>Created on 2019-11-19 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>


The package comes with nice functionality for:
- working with [windows](https://tsibble.tidyverts.org/articles/window.html) (actually, looking into it, package [`slide`](https://davisvaughan.github.io/slide/index.html) looks like it's superseding `tsibble` in terms of working with windows), 
- working with [missing data](https://tsibble.tidyverts.org/articles/implicit-na.html)
- it's easy to [aggregate over calendar periods](to aggregate over calendar periods) which could be handy.

Here's an example of a [forecasting workflow](https://robjhyndman.com/hyndsight/fable/) using `tsibble`  and `fable`.

Potentially we could try out a couple formats and set up some simple benchmarks to decide? Happy to hear your thoughts.

**Comment by pfistfl** [comment]
Hey @annakrystalli , nice to hear from you. 

I also like `tsibble`.

I think one thing we should aim for though is to stay light on dependencies.
From its DESCRIPTION:

```
Imports: 
    anytime (>= 0.3.1),
    dplyr (>= 0.8.1),
    lifecycle,
    lubridate (>= 1.7.0),
    purrr (>= 0.2.3),
    rlang (>= 0.2.0),
    tibble (>= 2.0.1),
    tidyselect,
    vctrs (>= 0.2.0)
```

So if we feel like we want to use `tsibble`, we have to find strong arguments for using all those dependencies.

**Comment by berndbischl** [comment]
using tsibble basically puts this package outside of the design principles of everything else in mlr3. i would rather avoid this

**Comment by pfistfl** [comment]
To add some thoughts to this:
I feel if `mlr3forecast` depends on `tsibble` **because** it has a lot of functionality we would otherwise have to rewrite and maintain, this is acceptable. 
You can still use the whole **mlr3verse** without all those dependencies, and only if you want to deal with time-series data, the dependencies are needed. 
I think this is one of the big benefits of splitting up into different packages.

Comments @mllg @berndbischl ?

**Comment by berndbischl** [comment]
I disagree. We have steered the project into a different direction and we cannot unilaterally now move away from this. This influences so many things, installation, configuration, documentation, that I dont think we can make a lighthearted decision here simply due to "easy of functionality". And because of the complexity of the topic, I also do not think that that can ever be discussed asynchronously with a few persons in an issue thread. Very certainly this cannot be dicussed without @mllg.
But people need to be able to make arguments about this and ask questions.

**Comment by SteveBronder** [comment]
I work with time series data on a daily basis and can say data.table is probably the way to go for storage

**Comment by berndbischl** [comment]
> I work with time series data on a daily basis and can say data.table is probably the way to go for storage

that would be absolutely preferable from a "mlr3 consistency" point of view. the problem is, dt does not give us too much in terms of service functionalitly specifically for ts. i am unsure how "suboptimal" it is. it would be good to write down a list of "super useful" operations we really need. and also evaluate their "difficuly" if we would have to reimplement them.

is that already in scope and being done?

**Comment by SteveBronder** [comment]
What sort of stuff do we need beyond rolling operations? Idt the storage needs that many bells and whistles

**Comment by berndbischl** [comment]
thats exactly my point, or what i want to propose to discuss. becaue then the decision is simplified a lot. 
OTOH i really cannot judge on the topic WHAT operations are needed. what i CAN tell you:

a) thinking about exactly this NOW is super important. a "type" is not defined by its storage format (only) but by its operations. and whether a type is approriate, is determined by which operations you need.

b) IF you are correct that you only need rolling operations, it might be dt. 

---> but only the team can make that decision by think about this.

**Comment by berndbischl** [comment]
> [`tsibble`](https://github.com/tidyverts/tsibble) which is a dependency of `tsbox`.

is this a typo or am i misreading this? tsbox does not depend on depend on tsiblle?

**Comment by annakrystalli** [comment]
Thanks for the input.

I appreciate the constraints of the project and I'm not tied to `tsibble`. Indeed what I find most useful is understanding what tool to use under which circumstance so the discussions are really useful. I also misread the `tsbox` dependencies and got the `imports` confused with `suggests`.  

With that, I'm going to ask a stupid question but that might be useful for the conversation:

@SteveBronder you mention unnecessary bells and whistles in `tsibble`. 
- What exactly are the additional bells and whistles?
- What do they cost?  (obviously additional dependencies but also memory? compute time?)
- What are the benefits of the bells and whistles? Perhaps then we can think of whether they are worth the cost once we have a better idea of workflow.

In any case, it sounds like a robust and easy to work with framework for creating windows will be required, so the `tsibble` and `slide` packages might still be a source of inspiration.  

Regarding windows, might we need to store metadata about them? eg if any transformations are applied at the window level?

**Comment by berndbischl** [comment]
> you mention unnecessary bells and whistles in `tsibble`.

it is basically the long list of dependencies - and sometimes the instability of these. 
that's a constraint/preference that the mlr3 core team decided upon, so it basically comes from us, not from steve.
you can read a little bit here :
https://github.com/mlr-org/mlr3 (under design principles)
(the rest might be easier to explain in chat / or a call, not an issue thread)
also pls note that steve said: "maybe we don't need these bells and whistles" and I read that as:
"tsibble provides a lot of (nice) extra functionality, but maybe don't need that"

so it's like 
"pro tsibble": lots of extra functionality, which we MAYBE do not need
"con tsibble": inconsistent design compared to the rest in mlr3.

 PS: 
a also looked at the API / cheat sheet of tsbox. it seems to provide EXACTLY what we want? 
they explicitly are "format agnostic" and say they support "plain df / dt" data. they also support casting from wide to long.....? isnt that really good?
NB: I haven't used the package, so I cannot judge implementation quality.
also: if we EVER WOULD decide to switch out our data format or backend (mlr3 kinda allows this anyway) we are now INDEPENDENT of that format if most of our operation come from tsbox?
(probably not 100% true, but 90% thruthiness would already be great)

**Comment by berndbischl** [comment]
also regarding the wide vs long format, which we have to decide for the task / backend:
assuming we would use a dt,  the long format is simply more general and flexible? i do agree with @SteveBronder that for most "ml stuff" that comes later we need a wide-format, but if we can easily convert, and conversion is simply 1 line of code, using the more general format might scale better into the future?

**Comment by berndbischl** [comment]
> * Does the current format need a specific DataBackend, or just a different Task Type?

i think thats an important question. i think you cannot answer this without writing down a 
prototypical "flow" of operations in pseudocode. i guess, especially for the long format, you need also a different backend, but i depends a lot on where you do certain format-conversion, and for that one need to see something to discuss it

**Comment by SteveBronder** [comment]
First off I def appreciate the input @annakrystalli! Working with time series is hard and weird! I was on the train earlier so sort of threw that comment their without elaborating

> What exactly are the additional bells and whistles?

Off the top of my head I think the data level bells and whistles we want to do everything else are

1. Rolling operations (running min, max, mean, correlations, etc.)
2. creating time slices (for things like cv)
3. (optional) imputing missing dates to make time series periodic

1 and 2 are pretty easy to 'roll' (and data.table has a bunch of the rolling stuff). (3) is pretty complex, like for instance certain stock markets can be opened or closed on certain days (think North American holidays and Indian holidays when working with forex data around easter). The user would probably want/need to specify what their data's periodicity is so we should probably push it to the discussion on data preprocessing.

> What do they cost? (obviously additional dependencies but also memory? compute time?)

`data.table` has [`rollfun`](https://github.com/Rdatatable/data.table/issues/2778) and friends, memory is going to be needed for the new columns, compute time is going to be pretty dependent on how the function is written, for the things data.table doesn't have there's also `RcppRoll` which is v fast

> What are the benefits of the bells and whistles? Perhaps then we can think of whether they are worth the cost once we have a better idea of workflow.

^I agree that what we do and don't need atm is a bit unclear. Personally I think it would be better to keep things lite then when we need stuff figure out if a package does it well enough to depend on or if it's worth rolling ourselves.

Also personally again, there's the sort of meta-issue that tsibble is in the tidyverse which is kind of notorious for making breaking changes. I use tidyverse stuff for reports etc. but for packages I'm wary of that

**Comment by SteveBronder** [comment]
> a also looked at the API / cheat sheet of tsbox. it seems to provide EXACTLY what we want?

I'm pro-tsbox, `anytime` is also pretty nice for handling dates. I think it has options to even handle nanosecond time periods

> Regarding windows, might we need to store metadata about them? eg if any transformations are applied at the window level?

metadata as in more than the actual slice size? I'm not sure but shooting from the hip I'd say there wouldn't be much. Like for N step ahead forecasting you need the previous state, but I think that would be handled somewhere else. Maybe you would need metadata for adjusting CV cuts for groups of time series (so for company A, B, and C March 1st 2019 is all in the training set and not half training half test). But I think that's a one time gig so we wouldn't need to store meta-info for that

Is the actual mlr3 task also part of this RFC or will that be discussed separately?

>  i do agree with @SteveBronder that for most "ml stuff" that comes later we need a wide-format, but if we can easily convert, and conversion is simply 1 line of code, using the more general format might scale better into the future?

I think it should be a one liner `cast()` to get from long to wide. I need to readup on how yinz handle data backends now

**Comment by berndbischl** [comment]
> I think it should be a one liner `cast()` to get from long to wide.

but ts_box also has ts_wide and ts_long? doesnt that do the casting for us?

**Comment by SteveBronder** [comment]
> but ts_box also has ts_wide and ts_long? doesnt that do the casting for us?

sure also fine

File: attic/data_backend_forecasting.md [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+```r
+DataBackendLong = R6Class("DataBackendLong",
+  inherit = mlr3::DataBackend,
+  cloneable = FALSE,
+
+  public = list(
+    id_col = NULL,
+    value_col = NULL,
+
+    initialize = function(data, primary_key, id_col) {
```
Comment (pfistfl) [inline attic/data_backend_forecasting.md]: The primary key is the time column, the id column is equivalent to the `id` column in the tsbox example above

**Pull request closed without merging**